// @mintlify/previewing/dist/local-preview/client.js
import { jsx as _jsx } from "react/jsx-runtime";
import fse from 'fs-extra';
import got from 'got';
import isOnline from 'is-online';
import { pipeline } from 'node:stream/promises';
import tar from 'tar';
import { DOT_MINTLIFY, DOT_MINTLIFY_LAST, VERSION_PATH, TAR_PATH, TARGET_MINT_VERSION_URL, } from '../constants.js';
import { addLog, clearLogs } from '../logging-state.js';
import { ErrorLog, SpinnerLog, WarningLog } from '../logs.js';
import { restoreMintlifyLast, getTarUrl } from '../util.js';
export const getTargetMintVersion = async () => {
    const hasInternet = await isOnline();
    if (!hasInternet) {
        return undefined;
    }
    try {
        const response = await got(TARGET_MINT_VERSION_URL);
        return response.body;
    }
    catch (error) {
        return undefined;
    }
};
export const downloadTargetMint = async ({ targetVersion, existingVersion, }) => {
    if (fse.existsSync(DOT_MINTLIFY)) {
        fse.moveSync(DOT_MINTLIFY, DOT_MINTLIFY_LAST, { overwrite: true });
    }
    fse.ensureDirSync(DOT_MINTLIFY);
    clearLogs();
    addLog(_jsx(SpinnerLog, { message: "installing mintlify framework..." }));
    const tarUrl = getTarUrl(targetVersion);
    let currentVersion = targetVersion.trim();
    try {
        await pipeline(got.stream(tarUrl), fse.createWriteStream(TAR_PATH));
    }
    catch (error) {
        if (existingVersion) {
            clearLogs();
            addLog(_jsx(WarningLog, { message: `failed to install mintlify framework version ${currentVersion}, ${error}, falling back to existing version: ${existingVersion}` }));
            currentVersion = existingVersion;
            restoreMintlifyLast();
            return;
        }
        else {
            clearLogs();
            addLog(_jsx(ErrorLog, { message: `failed to install mintlify framework version ${currentVersion}, ${error}` }));
            process.exit(1);
        }
    }
    try {
        clearLogs();
        addLog(_jsx(SpinnerLog, { message: "extracting mintlify framework..." }));
        tar.x({
            sync: true,
            file: TAR_PATH,
            cwd: DOT_MINTLIFY,
            onwarn: (_code, message, _data) => {
                throw new Error(message);
            },
        });
    }
    catch (error) {
        if (existingVersion) {
            clearLogs();
            addLog(_jsx(WarningLog, { message: `failed to extract mintlify framework version ${currentVersion}, ${error}, using existing version: ${existingVersion}` }));
            currentVersion = existingVersion;
            restoreMintlifyLast();
            return;
        }
        else {
            clearLogs();
            addLog(_jsx(ErrorLog, { message: `failed to extract mintlify framework version ${currentVersion}, ${error}` }));
            process.exit(1);
        }
    }
    clearLogs();
    fse.removeSync(TAR_PATH);
    if (fse.existsSync(DOT_MINTLIFY_LAST)) {
        fse.removeSync(DOT_MINTLIFY_LAST);
    }
    fse.writeFileSync(VERSION_PATH, currentVersion);
};
